Added separate_job experiment option to batch experiments

yaelbh opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

One half of #920.
When set to True, circuits coming from different sub-experiments are routed to different jobs. Default value is False.

nkanazawa1989 commented on qiskit_experiments/framework/composite/batch_experiment.py:
```diff
@@ -127,3 +127,69 @@ def _remap_qubits(self, circuit, qubit_mapping):
+                if not hasattr(circuits[0], "local_metadata"):
+                    is_first_time = True
+                    for circ in circuits:
+                        circ.local_metadata = circ.metadata
```
I don't think dynamically adding attribute like this is right approach. For example, if we decide to set `__slots__` to `QuantumCircuit`, this will no longer work. How local_metadata attribute is consumed by the job? Seems like this is only used to sort flatten circuits by sub experiment index. I think you can separately store list of metadata, and do something like
```python
circs_by_subexps = {}
for circ, metadata in zip(circuits, metadata_list):
    ...
    # sort by component index, add truncated metadata
```

nkanazawa1989 commented on qiskit_experiments/framework/composite/batch_experiment.py:
```diff
@@ -127,3 +127,69 @@ def _remap_qubits(self, circuit, qubit_mapping):
+                    # For batch experiments the composite metadata is always a list of length 1,
+                    # because unlike parallel experiment, each circuit originates from a single
+                    # sub-experiment.
+                    circ.local_metadata = circ.local_metadata["composite_metadata"][0]
```
```suggestion
                    circ.metadata = circ.local_metadata["composite_metadata"][0]
```
why this updates local_metadata? IBM provider checks `local_metadata` attribute by any chance?

yaelbh replied:
You can't override `metadata` because the analysis requires the full, non-truncated metadata. This is why we create `local_metadata`. The provider is not related, no one except for us is aware of `local_metadata`.

nkanazawa1989 commented on qiskit_experiments/framework/composite/batch_experiment.py:
```diff
@@ -127,3 +127,69 @@ def _remap_qubits(self, circuit, qubit_mapping):
+                    # because unlike parallel experiment, each circuit originates from a single
+                    # sub-experiment.
+                    circ_index = circ.local_metadata["composite_index"][0]
+                    if circ_index in circs_by_subexps:
```
You can also use [defaultdict](https://docs.python.org/3/library/collections.html#collections.defaultdict) if you want.

yaelbh commented on qiskit_experiments/framework/composite/composite_experiment.py:
```diff
@@ -175,6 +179,14 @@ def _finalize(self):
+                "separate_jobs", False
+            ) and subexp.experiment_options.get("separate_jobs", False):
+                raise QiskitError(
+                    "It is not allowed to equest to separate jobs in a child experiment,"
```
Then the inner batch will be done in one job (up to automatic splitting due to size)

yaelbh commented:
@nkanazawa1989 Thanks a lot for the review. CI is not passing because of tox issues. But I addressed your comments and will appreciate if you can re-review.

@haggaila You can try the current version (with the fixes following Naoki's review) on your use case.

nkanazawa1989 reviewed (approved):
Thanks @yaelbh this looks good to me.
